src/cpp/rtps/writer/StatefulWriter.cpp

```diff
@@ -643,8 +643,25 @@ void StatefulWriter::deliver_sample_to_intraprocesses(
             // send it a personal GAP.
             if (SequenceNumber_t::unknown() != gap_seq)
             {
-                intraprocess_gap(remoteReader, gap_seq, change->sequenceNumber);
-                remoteReader->acked_changes_set(change->sequenceNumber);
+                // Verify the calculated gap_seq in ReaderProxy is a real hole in the history.
+                SequenceNumber_t min_seq = get_seq_num_min();
+                if (gap_seq < min_seq) // Several samples of the hole are not really already available.
+                {
+                    if (min_seq < change->sequenceNumber)
+                    {
+                        gap_seq = min_seq;
+                    }
+                    else
+                    {
+                        gap_seq = SequenceNumber_t::unknown();
+                    }
+                }
```

Comment:
This code is duplicated. I think we could put this logic inside `ReaderProxy::change_is_unsent`. It would receive the minimum seq of the history, as `requested_changes_set` is doing.

Reply:
Done in d1e6bbb14